// yandex/web/Marker.js
import React, { Component } from 'react'
import compose              from 'recompose/compose'
import defaultProps         from 'recompose/defaultProps'
import lifecycle            from 'recompose/lifecycle'
import withHandlers         from 'recompose/withHandlers'
import withStateHandlers    from 'recompose/withStateHandlers'


const enhance = compose(
  defaultProps({
    iconColor: `#b69500`,
    iconImageHref: 'https://i.imgur.com/hetF82H.png',
    position: [55.76, 37.57],
  }),
  withStateHandlers(() => ({
    instance: null,
  }), {
    instanceSet: () => (instance) => ({ instance }),
  }),
  withHandlers({
    mount: ({
      iconColor,
      instanceSet,
      iconImageHref,
      map,
      position,
      ymaps,
    }) => () => {
      const instance = new ymaps.Placemark(position, {
        balloonContent: `xxx`
      }, {
        iconLayout: 'default#image',
        iconImageSize: [42, 42],
        iconImageHref,
        iconImageOffset: [-20, -42],
        iconColor,
      })

      map.geoObjects.add(instance)

      instanceSet(instance)
    },
    unmount: ({ instance, map }) => () => {
      map.geoObjects.remove(instance)
    }
  }),
  lifecycle({
    componentDidMount() {
      this.props.mount()
    },
    componentWillUnmount() {
      this.props.unmount()
    }
  })
)

const Marker = enhance(() => null)

export default Marker
